### ISSUE TITLE
Clarify API on accepting styles

### ISSUE THREAD

**Issue by adnasa** [vutran/omnibar#4]
One way to do this is add an actual API exposed to the consumer here.

```js
<Omnibar
     getRowStyle = () => {}
     getListStyle = () => {}
     getRootStyle = () => {}
     getInputStyle = () => {}
/>

// for row style
getRowStyle = ({ isFocused }) => {

}

// for list style
getListStyle = ({ isOpen }) => {}

// for the wrapping component
getRootStyle = ({ isLoadingSearchResults, isOpen }) => {}

// for the input
getInputStyle = ({ isFocused /*, isInvalid */ }) => {}
```
Now this is not much better that just plain `style` object, but this can offer some advantage. the naming is clear; row, list, input, and root (we can rename this to `container`). Also this could induce memoization on the consumer side which could be benefitial

**Comment by adnasa** [comment]
This is if we want to continue with using `style` prop.
I don't see any reason why we wouldn't go over to using classNames, then it would be agnostic to the stack used on the consumer side


```
import classnames from 'classnames',
import styles from './magic.css'

<Omnibar
     getRowClassName={({ isFocused }) => {}},
     getListClassName={({ isOpen }) => {}},
     inputClassName,
     containerClassName,
/>
```

Maybe this could pave way to have props such as `height` and `width` removed, since they are mixing concerns

**Comment by vutran** [comment]
I like this API design. Very similar to the render prop function used in [paypal/downshift](https://github.com/paypal/downshift). I've thought about wrapping `downshift` inside of `omnibar` as well but haven't had time instead of writing a competing framework.

My only gripe is that `downshift` is too custom and my goals with `omnibar` is to make it just a drop-in with minimal code setup but allowing flexibility if needed.

Thoughts?

**Comment by adnasa** [comment]
downshift could be something worth looking into, but I'm not sure if we need to go that far. 
It seems to me that it would be pretty excessive to integrate that for now.

my train of thought is that we add a clean API ourselves, this shouldn't be too hard to do (from the top of my head)

I'll look into react-select on this and see what learning we can get from there...
will get back to you 😉

**Comment by adnasa** [comment]
ohh,
I must've forgotten about this part, but `react-select` suggest that the one imports their css into your application. ugh 🤕

The real question now is if this is a good trade-off.
hmmmmmmm... here is my list of blabber

approach | pro | con
--- | --- |---
include "default" css on build for app | all responsibility and "decision" on the styling is determined on the implementation side. the API we write for custom styling can be even more simplistic, as long as the API is __rich__ | obviously it doesn't look good if they don't do this, but this is hardly a bad thing.
down-shift | All rendering responsibility on the implementation side (styles, other props) and possibility with custom formatting of the output. it'll almost fulfill the idea of compound component composition | ???

I like the down-shift approach now 🤣
let's try it out!!

**Comment by vutran** [comment]
Yeah, downshift may be better. I hate having to include an external css in my build. I just want to import my component module and it should look styled by default.

**Comment by vutran** [comment]
One note: for styling items, would make sense to move that responsibility to the `resultRenderer` to avoid compounding the `<Omnibar/>` component from too many props.

```typescript
interface Item {
  title: string;
  url: string;
}

interface Props {
  item: Item;
  isFocused: boolean;
  isHighlighted: boolean;
}

export default function ResultItem(props: Props) {

  // apply styles here based on props.isFocused, props.isHighlighted, etc...

  return (
    <div><a href={item.url}>{item.title}</a></div>
  );
}
```

**Comment by adnasa** [comment]
I haven't jumped into trying to integrate downshift just yet,

I started with [an approach](https://github.com/vutran/omnibar/compare/master...adnasa:aa-get-style-method) (clean up API) and it already made a bit of a difference

I'm not entirely happy about this atm, but..

👍  now the consumer can return contextual styles, based on state of input and item (row).

👍  A personal opinion; this has also paved way to have unecessary props removed (such as height, width, maxHeight). I'm wondering what value these props provide. Care to shed light on this?

👍  I also found out that the responsibility of setting the "highlight" state on the item is shared between `Omnibar` and `ResultsItem`. We can think about how we deal with this on #13
